ui/src/containers/UndoRedo.tsx

```diff
@@ -0,0 +1,60 @@
+import RedoIcon from "@mui/icons-material/Redo";
+import UndoIcon from "@mui/icons-material/Undo";
+import { Box, Button } from "@mui/material";
+import { useUrl } from "hooks";
+import { connect, ConnectedProps } from "react-redux";
+import { Link as RouterLink } from "react-router-dom";
+import { ActionCreators as UndoActionCreators } from "redux-undo";
+import { RootState } from "rootReducer";
+import { AppDispatch } from "store";
+
+const mapStateToProps = (state: RootState) => {
+  return {
+    canUndo: state.url.past.length > 0,
+    canRedo: state.url.future.length > 0,
+  };
+};
+
+const mapDispatchToProps = (dispatch: AppDispatch) => {
+  return {
+    onUndo: () => dispatch(UndoActionCreators.undo()),
+    onRedo: () => dispatch(UndoActionCreators.redo()),
+  };
+};
+
+const connector = connect(mapStateToProps, mapDispatchToProps);
+
+type propsFromRedux = ConnectedProps<typeof connector>;
+
+function UndoRedo({ canUndo, canRedo, onUndo, onRedo }: propsFromRedux) {
+  const urlHistory = useUrl();
+
+  return (
+    <Box>
+      <Button
+        onClick={onUndo}
+        disabled={!canUndo}
+        component={RouterLink}
+        to={urlHistory.present}
```

Comment:
It seems like `urlHistory.present` is correct because this is the URL before the state change and subsequent navigation occurs, so this is where we would like to navigate to when we undo a state change.